aips/aip-116.md

```diff
@@ -326,47 +395,63 @@ A new npm package named `@aptos-labs/siwa` will be created to house utilities fo
 Below are the primary functions that will be provided as part of the `@aptos-labs/siwa` utilities:
 
 ```ts
+export type VerificationResult<T> =
+  | { valid: true; data: T }
+  | { valid: false; errors: string[] };
+
 /**
- * Create a SIWA plain text message from the `AptosSignInInput` (including required fields) following the ABNF format defined in the Sign in with Aptos
+ * Create a SIWA Message from the `AptosSignInInput` with required fields following the ABNF format defined in the Sign in with Aptos
  * specifications.
  *
- * @param input The input fields to create the SIWA plain text message from.
+ * @param input The `AptosSignInInput` with required fields to create the SIWA message from.
  *
- * @returns The SIWA plain text message.
+ * @returns The SIWA message.
  */
-function createSignInMessagePlainText(
+function createSignInMessageText(
   input: AptosSignInInput & AptosSignInRequiredFields
 ): string;
 
 /**
- * Verifies a SIWA plain text message against expected `AptosSignInInput` fields (including required fields).
+ * Parse a SIWA message into an `AptosSignInInput` object with the required fields.
  *
- * @param message The SIWA plain text message to verify.
- * @param expected The expected fields to verify against the input.
+ * @param text The SIWA message to parse.
  *
- * @returns The `AptosSignInInput` fields that were verified.
+ * @returns The parsed `AptosSignInInput` object with the required fields.
  */
-function verifySignInMessage(
-  message: string,
-  expected: AptosSignInInput & AptosSignInRequiredFields
+function parseSignInMessageText(
+  text: string
 ): VerificationResult<AptosSignInInput & AptosSignInRequiredFields>;
 
 /**
- * Verifies certain components of a `AptosSignInOutput` from a `signIn` method against `AptosSignInInput` fields.
+ * Using the `publicKey` and `signature`, verify that the `signature` is valid for the `message`.
+ *
+ * @param output The `AptosSignInOutput` to verify against the input.
+ *
+ * @returns The `AptosSignInInput` fields that are parsed from the message.
+ */
+function verifySignInSignature(output: {
+  publicKey: PublicKey;
+  signature: Signature;
+  message: string;
+}): VerificationResult<AptosSignInInput & AptosSignInRequiredFields>;
```

Comment:
This is rather odd to me: typically signature verification returns `true` / `false`.

Furthermore, deserializing during signature verification is an anti-pattern. It's more secure to verify a signature over a structured object by (1) serializing it and (2) checking the signature on the serialization (+ whatever domain separation is needed).